### PR TITLE
🐛🐛 Raise original error only when !predicate

### PR DESCRIPTION
**Pull request by louib**
bringing back the `RetryError` that was removed in https://github.com/unitoio/backoff-decorator/pull/11.

We need to throw a `RetryError` when the max retries is reached and the error should've been retried. The original error is only thrown if it doesn't match the predicate (an "unexpected" error).

**We will need to update the dashboards (especially babylon) after this is merged.**

### CODE REVIEW

File: src/backoff.ts [hunk 1]
```diff
@@ -105,16 +109,21 @@ export async function retry(
       return result;
     } catch (err) {
       originalError = err;
-      if (!predicate || !predicate(err, args)) {
-        throw err;
+      if (!predicate || !predicate(originalError, args)) {
```
Comment (erykwarren) [inline src/backoff.ts]: not sure of the use of the `originalError` variable. Seems that it is simply equivalent to `err`...

File: src/backoff.ts [hunk 2]
```diff
@@ -105,16 +109,21 @@ export async function retry(
...
 
     numRetries++;
     if (numRetries > maxRetries) {
       // The last error that was raised before the max retries is reached
-      // will be the one thrown to the end-user.
-      throw originalError;
+      // will be the one thrown to the end-user, unless it's a "known" error
+      // that was configured for backoff.
+      if (!predicate || !predicate(originalError, args)) {
+        throw originalError;
```
Comment (erykwarren) [inline src/backoff.ts]: I don't understand this block. It should already be handled by the `if` on line 112 no?

**Review by erykwarren — CHANGES_REQUESTED** [review]

File: src/backoff.ts (hunk 1 above)
Comment (louib) [reply src/backoff.ts]: it is, but I wanted both if conditions to be written the same way...


\#OCD

File: test/backoff_test.ts [hunk 3]
```diff
@@ -192,12 +207,12 @@ describe('backoff', () => {
       }
     });
 
-    it('throws an error if maxRetries is reached', async () => {
+    it('throws a RetryError if maxRetries is reached', async () => {
```
Comment (erykwarren) [inline test/backoff_test.ts]: This test is a double of the one on line 143!

**Review by erykwarren — APPROVED** [review]

File: src/backoff.ts [hunk 4]
```diff
@@ -1,3 +1,7 @@
+export class RetryError extends Error {
+  name: 'RetryError';
```
Comment (ronjouch) [inline src/backoff.ts:2]: @louib this isn't enough to get the stringify behavior you expect, you need

```js
class RetryError extends Error {
  constructor(message) {
    super(message);
    this.name = "RetryError";
    // Or
    this.name = this.constructor.name
  }
}
```

That's what we do in BaseError.

**Review by ronjouch — COMMENTED** [review]
Does that fix yesterday weirdness? Can you show the current code path and how that fixes it?

**Pull request closed without merging**